static/js/table-of-contents.js

```diff
@@ -0,0 +1,35 @@
+function clearActiveStatesInTableOfContents() {
+	document.querySelectorAll('#TableOfContents ul li a').forEach((section) => {
+		section.classList.remove('active');
+	});
+}
+
+window.addEventListener('DOMContentLoaded', () => {
+
+	const observer = new IntersectionObserver(entries => {
+		var currentIR = 0;
+		entries.forEach(entry => {
+			const id = entry.target.getAttribute('id');
+			if (entry.intersectionRatio > currentIR) {
+				currentIR = entry.intersectionRatio;
+				clearActiveStatesInTableOfContents();
+				document.querySelector(`#TableOfContents ul li a[href="#${id}"]`).classList.add('active');
+			}
+		});
+
+	});
+
+	// Track all headers that have an `id` applied
+	document.querySelectorAll('article h3[id], article h2[id]').forEach((section) => {
+		observer.observe(section);
+	});
+});
+
+// Post title doesn't get parsed as it is not part of the content file
+(function addPostTitle2ToC() {
+	var toc = document.getElementById('TableOfContents');
+	if (!toc) return;
+	var title = document.getElementById("post__title").innerText;
+	var ul = toc.querySelector('ul');
+	ul.outerHTML = `<ul><li><a href="#post__title" class="active">${title}</a>${ul.outerHTML}</li></ul>`;
```

Comment:
Not sure about this guy. Is this considered bad practice in web development? It feels like a dirty patch, but It's the easier solution I could come up with.

Reply:
Is it common to have the main title in a table
Of contents?

Reply:
It is in books, at least. It usually has some kind of header, even if it is "Table of contents"

Reply:
Hmmm.

If the h1 belongs in the ToC maybe it would be best to find a solution for rendering the title with the rest of the post so that it's targetable. Is that the main issue here?

Reply:
Yes! In principle that would solve the issue.

Reply:
> Is it common to have the main title in a table of contents?
I actually don't like the main title in the TOC either but luckily [it can be configured](https://gohugo.io/getting-started/configuration-markup/#table-of-contents). Just take care to explain this in the README is good enough I guess.

Reply:
@jakewies oops I know why this is happening. It's because the page is using `<h1>` as title:
![image](https://user-images.githubusercontent.com/7024160/85605186-8ade8a80-b684-11ea-9ef3-94c42f13bb20.png)

The page's title should probably use some specific styling instead of using `<h1>`, what do you think?

Reply:
Semantically, I feel like the title of the page has to be h1. We would need to find a way to render it with the rest of the post to make it targetable by the ToC, as @jakewies suggested above

Reply:
oh right, my bad. I have a look at other themes and it looks like they're using `<h1>` for titles too. But yeah changing the `config` should be good enough to hide/show the title as needed.

Reply:
Sorry just catching up on this PR. Just to recap:

- the title of the blog post _should_ be an `h1`
- We can add configuration to hide/show the `h1` in the ToC
- BUT, ToC logic does not have access to the `h1` because it is not being rendered as part of the blog post content. It's being rendered separately via frontmatter.

So our options are to use JS to target the title of a blog post for ToC, or find some way to render it with the rest of the content so Hugo's built in ToC can take care of it. The problem is I don't know how you would do the latter unless you forced the user to explicitly write their post title in their markdown _outside_ of the frontmatter. Not sure I like that.

Reply:
oh after reading the JS code more carefully I understand what is doing now. So the `addPostTitle2ToC()` function was adding the title to the TOC. 

Personally I wouldn't want the title in the TOC as it seems repetitive, especially when the user is at the top of the page where the real title is also visible. But we can definitely make it an option to run this script if the user wants the title in the TOC. Not to mention @xvallspl's hack is nice too, I can't think of another way.

However, is it possible to give the title in TOC another CSS class, like `toc_title` instead of `active`, in case users want it to look different than the rest of the TOC?

Reply:
> Personally I wouldn't want the title in the TOC as it seems repetitive, especially when the user is at the top of the page where the real title is also visible.

I tend to agree.

> However, is it possible to give the title in TOC another CSS class, like toc_title instead of active, in case users want it to look different than the rest of the TOC?

Actually the title will never receive the `"active"` class. It's just not possible with the way the JS listener functions.

Reply:
> Actually the title will never receive the "active" class.

Oh i was talking about this hardcoded `"active"` class when we insert the title into the TOC:
```
ul.outerHTML = `<ul><li><a href="#post__title" class="active">${title}</a>...
```

Reply:
Oh I see. When the page loads it gets the `active` class now I see it.